### PR TITLE
 [browser][debugger][MT] Make operations on dictionary thread-safe.

### LINKED ISSUE #87315: [browser][debugger][MT] CallFunctionOnTests.RunOnInvalidCfoId
```
[xUnit.net 00:08:09.53]     DebuggerTests.CallFunctionOnTests.RunOnInvalidCfoId(eval_fn: "negative_cfo_test ();", bp_loc: "/other.js", line: 64, col: 1, use_cfo: False) [FAIL]
Detected a container, disabling sandboxing for debugger tests.
  Failed DebuggerTests.CallFunctionOnTests.RunOnInvalidCfoId(eval_fn: "negative_cfo_test ();", bp_loc: "/other.js", line: 64, col: 1, use_cfo: False) [1 ms]
  Error Message:
   System.AggregateException : One or more errors occurred. (Index was outside the bounds of the array.)
---- System.IndexOutOfRangeException : Index was outside the bounds of the array.
  Stack Trace:
     at DebuggerTests.Inspector.OpenSessionAsync(Func`3 getInitCmds, String urlToInspect, TimeSpan span) in /_/src/mono/wasm/debugger/DebuggerTestSuite/Inspector.cs:line 377
   at DebuggerTests.DebuggerTestBase.InitializeAsync() in /_/src/mono/wasm/debugger/DebuggerTestSuite/DebuggerTestBase.cs:line 173
----- Inner Stack Trace -----
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at System.Collections.Generic.Dictionary`2.set_Item(TKey key, TValue value)
   at DebuggerTests.InspectorClient.SendCommand(SessionId sessionId, String method, JObject args, CancellationToken token) in /_/src/mono/wasm/debugger/DebuggerTestSuite/InspectorClient.cs:line 98
   at DebuggerTests.Inspector.OnMessage(String sessionId, String method, JObject args, CancellationToken token) in /_/src/mono/wasm/debugger/DebuggerTestSuite/Inspector.cs:line 219
  Standard Output Messages:

...

[2023-06-08T20:02:59] Inspector-67 Error: _isFailingWithException. System.AggregateException: One or more errors occurred. (Index was outside the bounds of the array.)
 |                                            ---> System.IndexOutOfRangeException: Index was outside the bounds of the array.
 |                                              at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
 |                                              at System.Collections.Generic.Dictionary`2.set_Item(TKey key, TValue value)
 |                                              at DebuggerTests.InspectorClient.SendCommand(SessionId sessionId, String method, JObject args, CancellationToken token) in /_/src/mono/wasm/debugger/DebuggerTestSuite/InspectorClient.cs:line 98
 |                                              at DebuggerTests.Inspector.OnMessage(String sessionId, String method, JObject args, CancellationToken token) in /_/src/mono/wasm/debugger/DebuggerTestSuite/Inspector.cs:line 219
 |                                              --- End of inner exception stack trace ---
 |                                              at DebuggerTests.Inspector.OnMessage(String sessionId, String method, JObject args, CancellationToken token) in /_/src/mono/wasm/debugger/DebuggerTestSuite/Inspector.cs:line 220
 |                                              at DebuggerTests.Inspector.OnMessage(String sessionId, String method, JObject args, CancellationToken token) in /_/src/mono/wasm/debugger/DebuggerTestSuite/Inspector.cs:line 219
 |                                              at DebuggerTests.Inspector.OnMessage(String sessionId, String method, JObject args, CancellationToken token) in /_/src/mono/wasm/debugger/DebuggerTestSuite/Inspector.cs:line 220
 |                                              at DebuggerTests.Inspector.OnMessage(String sessionId, String method, JObject args, CancellationToken token) in /_/src/mono/wasm/debugger/DebuggerTestSuite/Inspector.cs:line 223
 |                                              at DebuggerTests.Inspector.OnMessage(String sessionId, String method, JObject args, CancellationToken token) in /_/src/mono/wasm/debugger/DebuggerTestSuite/Inspector.cs:line 221
 |                                              at DebuggerTests.Inspector.<OpenSessionAsync>b__40_1() in /_/src/mono/wasm/debugger/DebuggerTestSuite/Inspector.cs:line 362.
```

[Build](https://dev.azure.com/dnceng-public/public/_build/results?buildId=300315&view=logs&jobId=2112d60c-a661-5638-18dd-f51355a8b038)
[log](https://helixre107v0xdcypoyl9e7f.blob.core.windows.net/dotnet-runtime-refs-pull-86759-merge-33368a48c0f84a42ac/chrome-DebuggerTests.CallFunctionOnTests/1/console.8a5f1d20.log?helixlogtype=result)

### PR DESCRIPTION
**Pull request by ilonatommy**
Fixes https://github.com/dotnet/runtime/issues/87315.

`ConcurrentDictionary` should be thread-safe, [doc](https://learn.microsoft.com/en-us/dotnet/api/system.collections.generic.dictionary-2?view=net-7.0#thread-safety).

### CODE REVIEW

File: src/mono/wasm/debugger/DebuggerTestSuite/InspectorClient.cs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 
 using System;
 using System.Collections.Generic;
```
Comment (ilonatommy) [inline src/mono/wasm/debugger/DebuggerTestSuite/InspectorClient.cs:5]: ToDo: Remove. This namespace is not used anymore.

**Comment by ilonatommy** [comment]
CI passed but the failures are flaky so I will run again to make sure.

**Comment by ilonatommy** [comment]
Single threaded dbg failure is https://github.com/dotnet/runtime/issues/87053

**Review by maraf — APPROVED** [review]
Looks good to me 👍

File: src/mono/wasm/debugger/DebuggerTestSuite/InspectorClient.cs [hunk 2]
```diff
@@ -17,7 +17,7 @@ namespace DebuggerTests
 {
     internal class InspectorClient : DevToolsClient
     {
-        protected Dictionary<MessageId, TaskCompletionSource<Result>> pending_cmds = new Dictionary<MessageId, TaskCompletionSource<Result>>();
+        protected ConcurrentDictionary<MessageId, TaskCompletionSource<Result>> pending_cmds = new ConcurrentDictionary<MessageId, TaskCompletionSource<Result>>();
```
Comment (radical) [inline src/mono/wasm/debugger/DebuggerTestSuite/InspectorClient.cs:20]:
```suggestion
        protected readonly ConcurrentDictionary<MessageId, TaskCompletionSource<Result>> pending_cmds = new();
```

**Review by radical — APPROVED** [review]

File: src/mono/wasm/debugger/DebuggerTestSuite/InspectorClient.cs [hunk 3]
```diff
@@ -95,7 +95,13 @@ public virtual Task<Result> SendCommand(SessionId sessionId, string method, JObj
             if (sessionId != SessionId.Null)
                 o.Add("sessionId", sessionId.sessionId);
             var tcs = new TaskCompletionSource<Result>();
-            pending_cmds[new MessageId(sessionId.sessionId, id)] = tcs;
+
+            MessageId msgId = new MessageId(sessionId.sessionId, id);
+            pending_cmds.AddOrUpdate(msgId, tcs, (key, oldValue) => {
+                if (oldValue != tcs)
+                    logger.LogDebug($"Updating the pending command, key = {key}, oldValue= {oldValue}, newValue = {tcs}");
```
[the remark this answers is not in the document]
Comment (ilonatommy) [reply src/mono/wasm/debugger/DebuggerTestSuite/InspectorClient.cs:102]: No, the issue was adding the value when another thread already had a lock on that dictionary. I actually am not sure if we should inform about an update when two threads want to write into the same key.

[the remark this answers is not in the document]
Comment (radical) [reply src/mono/wasm/debugger/DebuggerTestSuite/InspectorClient.cs:102]: I would drop the message for now. And if/when we see a case where two threads are trying to write for the same msgId, then we can investigate further.

**Pull request merged**